examples/step-77/step-77.cc

```diff
@@ -0,0 +1,660 @@
+  // TimerOutput object to measure how much time every operation on each mesh
+  // costs, and reset the timer at the beginning of each cycle.
+  //
+  // As discussed in the introduction, it is not necessary to solve problems on
+  // coarse meshes particularly accurately since these will only solve as
+  // starting guesses for the next mesh. As a consequence, we will use a target
+  // tolerance of
+  // $\tau=10^{-3} \frac{1}{10^k}$ for the $k$th mesh refinement cycle.
+  //
+  // All of this is encoded in the first part of this function:
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::run()
+  {
+    GridGenerator::hyper_ball(triangulation);
+    triangulation.refine_global(2);
+
+    setup_system(/*initial_step=*/true);
+    set_boundary_values();
+
+    for (unsigned int refinement_cycle = 0; refinement_cycle < 6;
+         ++refinement_cycle)
+      {
+        computing_timer.reset();
+        std::cout << "Mesh refinement step " << refinement_cycle << std::endl;
+
+        if (refinement_cycle != 0)
+          refine_mesh();
+
+        const double target_tolerance = 1e-3 * std::pow(0.1, refinement_cycle);
+        std::cout << "  Target_tolerance: " << target_tolerance << std::endl
+                  << std::endl;
+
+        // This is where the fun starts. At the top we create the KINSOL solver
+        // object and feed it with an object that encodes a number of additional
+        // specifics (of which we only change the nonlinear tolerance we want to
+        // reach; but you might want to look into what other members the
+        // SUNDIALS::KINSOL::AdditionalData class has and play with them).
+        {
+          SUNDIALS::KINSOL<Vector<double>>::AdditionalData additional_data;
+          additional_data.function_tolerance = target_tolerance;
+
+          SUNDIALS::KINSOL<Vector<double>> nonlinear_solver(additional_data);
+
+          // Then we have to describe the operations that were already mentioned
+          // in the introduction. In essence, we have to teach KINSOL how to (i)
+          // resize a vector to the correct size, (ii) compute the residual
+          // vector, (iii) compute the Jacobian matrix (during which we also
+          // compute its factorization), and (iv) solve a linear system with the
+          // Jacobian.
+          //
+          // All four of these operations are represented by member variables of
+          // the SUNDIALS::KINSOL class that are of type `std::function`, i.e.,
+          // they are objects to which we can assign a pointer to a function or,
+          // as we do here, a "lambda function" that takes the appropriate
+          // arguments and returns the appropriate information. By convention,
+          // KINSOL wants that functions doing something nontrivial return an
+          // integer where zero indicates success. It turns out that we can do
+          // all of this in just 25 lines of code.
```

Comment:
This might be the first time that some novice C++ programmers see a lambda function. Might it be worthwhile writing one or two sentences that explain the syntax (especially the capture by reference, and why you want to do that)? Feel free to disagree :-)

Reply:
We actually use lambdas in a number of places -- see step-12, 13, 16. But it doesn't matter, some things are worth repeating, so I'll add a short snippet.

Reply:
Sorry for the noise then. It's been a long time since I've looked over those tutorials. Thanks for considering the suggestion anyway!